File: driveway_check_web/src/lib/identify/median-finder.ts
/** Finds grass medians inside a divided drive, and bridges them shut.
 *
 * Ported from Part1_Identify/Core/MedianFinder.cs.
 *
 * Morphological closing: buffer out then back in by the same distance. A gap
 * narrower than twice the reach closes and reappears as a difference against the
 * original; anything wider stays open. Two strips either side of a median are one
 * drive conceptually, so the median is bridged into the analysis polygon before
 * the skeleton runs — otherwise the shape reads as two separate drives. */

import { mitre_buffer, type JtsGeometry, type JtsPolygon } from "./jts";

/** Numerical noise floor, in square feet — NOT a judgement about how small a
 * median can be. If the buffer ever leaves larger slivers, fix the input
 * precision rather than raising this; raising it starts discarding real
 * geometry. */
const MIN_GAP_AREA = 0.5;

export function find_medians(polygon: JtsPolygon): JtsPolygon[]
{
	// Area over perimeter is half the width of an elongated region — derived, not
	// fitted, so it scales with the drive instead of being tuned to one plan.
	const reach = polygon.getArea() / polygon.getLength();
	if (!(reach > 0)) return [];

	// Mitre joins are load-bearing. Round joins produced seven false regions of
	// 4-21 sq ft at concave corners across the fixtures; mitre produced none, with
	// the one real median stable at 227 sq ft across 14 footprints x 8 rotations.
	const closed = mitre_buffer(mitre_buffer(polygon, reach), -reach);
	const filled = closed.difference(polygon);

	// A real median is ENCLOSED by paving: it is a void the drive wraps around, so
	// a thin collar drawn around it lies entirely on paving. A notch between two
	// bays is not - closing it draws a straight line across its mouth, and roughly
	// half that collar sticks out into open ground.
	//
	// Measured on the townhouse plan, which has no median at all: every one of the
	// 14 surviving candidates scores 0.47-0.66, against ~1.0 for a genuine void.
	// The cut sits at 0.9, i.e. "essentially fully enclosed" - a fraction, so it
	// does not scale with the plan, and it has ~0.25 of margin below it.
	//
	// This matters beyond the false positives: bridging these notches deforms the
	// drive enough that the degenerate-junction weld downstream stops firing, so
	// an over-eager median finder silently breaks segmentation too.
	const COLLAR_FT = 0.25;
	const ENCLOSED_FRACTION = 0.9;

	const found: JtsPolygon[] = [];
	for (let i = 0; i < filled.getNumGeometries(); i++)
	{
		const gap = filled.getGeometryN(i);
		if (gap.getGeometryType() !== "Polygon") continue;
		if (gap.isEmpty() || gap.getArea() < MIN_GAP_AREA) continue;

		const collar = mitre_buffer(gap, COLLAR_FT).difference(gap);
		const collar_area = collar.getArea();
		if (!(collar_area > 0)) continue;
		if (collar.intersection(polygon).getArea() / collar_area < ENCLOSED_FRACTION) continue;

		found.push(gap as JtsPolygon);
	}
	return found;
}

/** Unions the medians back in, so the skeleton sees one corridor. */
export function bridge(polygon: JtsPolygon, medians: readonly JtsPolygon[]): JtsPolygon
{
	let bridged: JtsGeometry = polygon;
	for (const median of medians) bridged = bridged.union(median);
	return bridged.getGeometryType() === "Polygon" ? (bridged as JtsPolygon) : polygon;
}

File: driveway_check_rhino/src/Part1_Identify/Core/MedianFinder.cs
using System.Collections.Generic;
using NetTopologySuite.Geometries;
using NetTopologySuite.Operation.Buffer;

namespace DrivewayChecker.Identify.Core
{
    // Finds landscape medians: the unpaved strip dividing one drive into two
    // travel lanes.
    //
    // A median is a void rather than paving, so nothing else in the pipeline
    // can see it -- and its absence breaks more than it appears to. On the
    // user's divided drive the two 8ft strips came out as two separate DRIVE
    // regions instead of one drive, AND the apron went undetected entirely:
    // the median's tip ends exactly where the apron mouth begins (measured
    // 0.002ft apart, which is correct design practice, not a drafting slip),
    // so the mouth crossed 2.13ft of grass and ApronFinder's SpansPaving
    // rejected it. Bridging the median fixes both at once, because everything
    // downstream then sees one continuous drive.
    //
    // What comes back is for ANALYSIS ONLY. Measurement and face tagging stay
    // on the real paving -- a median is not driveable surface and must never
    // enter a width or area check. See IdentifyRun.
    public static class MedianFinder
    {
        // Morphological closing: grow the paving by `reach`, then shrink it
        // back. Anything narrower than about 2*reach gets sealed shut on the
        // way out and does not reopen on the way back, so what survives the
        // round trip is exactly the set of gaps too narrow to be part of the
        // drive's own shape. Subtracting the original paving leaves the gaps
        // themselves.
        //
        // MITRE joins, not the default round ones, and this is the whole trick.
        // A round join cannot reproduce a sharp corner when it shrinks back, so
        // every concave corner in the footprint comes back as a small filled
        // wedge -- measured across the fixtures, seven false regions of 4 to 21
        // sq ft, which then need a size rule to tell them from a real median.
        // Mitre joins rebuild the corner exactly and those regions simply never
        // appear: across 14 footprints at 8 rotations each, the ONLY thing this
        // fills is the one real median, stable at 227 sq ft to the square foot
        // at every rotation.
        //
        // reach = area / perimeter is half the width of any elongated region,
        // so closing at that radius bridges precisely the gaps that are
        // narrower than the paving flanking them. That is the engineering
        // meaning of a median -- narrower than what it divides -- rather than a
        // fitted number, and it is why the gap inside a U-shaped drive is left
        // alone: on the staple fixtures the gap is 40ft between 13.3ft legs,
        // far too wide to seal.
        public static List<Polygon> Find(Polygon polygon)
        {
            var found = new List<Polygon>();

            double reach = polygon.Area / polygon.Length;
            if (reach <= 0) return found;

            var joins = new BufferParameters
            {
                JoinStyle = JoinStyle.Mitre,
                MitreLimit = 5.0,
                EndCapStyle = EndCapStyle.Flat,
            };
            var filled = polygon.Buffer(reach, joins).Buffer(-reach, joins).Difference(polygon);

            for (int i = 0; i < filled.NumGeometries; i++)
            {
                if (filled.GetGeometryN(i) is not Polygon gap) continue;
                // A convex footprint fills nothing, and NTS reports that as one
                // EMPTY polygon rather than as zero geometries. The area floor
                // is for slivers left by the buffer round trip's own arithmetic
                // -- it is numerical noise, far below any real dimension, not a
                // judgement about how small a median can be.
                if (gap.IsEmpty || gap.Area < 0.5) continue;
                found.Add(gap);
            }
            return found;
        }

        // The footprint with its medians bridged: what the skeleton, the runs
        // and the apron search all run against, so a divided drive reads as one
        // drive. The caller keeps the original for everything measurable.
        public static Polygon Bridge(Polygon polygon, IReadOnlyList<Polygon> medians)
        {
            Geometry bridged = polygon;
            foreach (var median in medians) bridged = bridged.Union(median);
            return bridged as Polygon ?? polygon;
        }
    }
}
